tests: add cmd/link, cmd/unlink integration test

eirinikos opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This change adds test coverage for `brew link` and `brew unlink`, though [these lines of `cmd/link.rb`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/link.rb#L53-L55) still lack coverage (I haven't yet figured out how the `LinkError` gets raised).

[These lines](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/link.rb#L53-L55) also lack coverage, though somehow they *were* tested when I tried using a combination of `test/testball.rb`+ `setup_test_formula "baz"` (instead of only `setup_test_formula "testball"`). Since there are three other `ensure` blocks that do the following, I was hoping that the `cmd`s could be refactored into their own method):
```
cmd("uninstall", "--force", testball)
cmd("cleanup", "--force", "--prune=all")
```

But I wanted to test a `keg_only` formula and I'm not sure if there's a way to modify the existing `test/testball` file to make it such.

MikeMcQuaid commented on Library/Homebrew/test/test_integration_cmds.rb:
```diff
@@ -668,4 +668,32 @@ def test_search
+
+    formula_file = setup_test_formula "testball"
+    cmd("install", "testball")
+    assert_match "Already linked: #{HOMEBREW_CELLAR}/testball/0.1",
```
As we're not testing that `install` works or doesn't in this function you can skip the `assert_match` here. `cmd` is sufficient as it'll verify the return code.

MikeMcQuaid commented on Library/Homebrew/test/test_integration_cmds.rb:
```diff
@@ -668,4 +668,32 @@ def test_search
+      cmd("link", "testball")
+    assert_match "Would remove", cmd("unlink", "--dry-run", "testball")
+
+    cmd("unlink", "testball")
```
think it's worth separating `unlink` into another test.

eirinikos commented:
Thanks @MikeMcQuaid and sorry for the late response!

I've removed the unnecessary `assert_match` and split `test_link_unlink` into `test_link` and `test_unlink`. Perhaps the split isn't as clean as can be, in which case I look forward to more feedback. 😄

UniqMartin commented:
I wonder if it would be feasibly possible to get rid of most of the `cmd` method calls just for setup and teardown (`install`, `uninstall`, `cleanup`). They are a bit expensive, often more fragile than a simple `rm_rf` (and other equivalent operations), and the point is testing `link` and `unlink`, not those other commands that are probably already covered somewhere else.

But this is just a gut feeling and I haven't looked more carefully into this, thus feel free to set me straight.

eirinikos commented:
> I wonder if it would be feasibly possible to get rid of most of the `cmd` method calls just for setup and teardown (`install`, `uninstall`, `cleanup`). They are a bit expensive, often more fragile than a simple `rm_rf` (and other equivalent operations), and the point is testing `link` and `unlink`, not those other commands that are probably already covered somewhere else.
> 
> But this is just a gut feeling and I haven't looked more carefully into this, thus feel free to set me straight.

Thanks @UniqMartin. I checked which files would be leaked without `cmd("uninstall")` and `cmd("cleanup")` in place. In addition to the files in the files in the Cellar, there are 4 other directory locations:

```
+ "/cache/testball-0.1.tbz",
+ "/locks/testball.brewing",
+ "/prefix/Library/LinkedKegs",
+ "/prefix/Library/LinkedKegs/testball",
+ "/prefix/opt",
+ "/prefix/opt/testball"
```

Would it be efficient to use `FileUtils.rm_rf` on all 5 different directories? There are 5-6 other integration tests that also use `cmd("uninstall")` and `cmd("cleanup")` in their `ensure` blocks; perhaps there's a way to also slim down those tests?

My new commit modifies `setup_test_formula` slightly to let it accept variations of `testball` (and other strings) as names, so that there's no need to `cmd("uninstall")` the original formula file. It shortens the test by 2 lines, but it adds another `cmd("uninstall")` to the `ensure` block. Not ideal, I know. 😞

MikeMcQuaid commented:
> Would it be efficient to use FileUtils.rm_rf on all 5 different directories?

Yep, I think so. I think favouring it over `uninstall` and `cleanup` is also worth doing 👍

eirinikos commented:
> Would it be efficient to use `FileUtils.rm_rf` on all 5 different directories?
> Yep, I think so. I think favouring it over `uninstall` and `cleanup` is also worth doing 👍

Thanks! I've updated the commit. If the usage of `FileUtils.rm_rf` looks OK, I figure I'll use it to replace the other tests that use `cmd("uninstall")` and `cmd("cleanup")` (in a separate pull request). :grin:

MikeMcQuaid commented on Library/Homebrew/test/test_integration_cmds.rb:
```diff
@@ -719,4 +719,43 @@ def test_services
+  ensure
+    FileUtils.rm_rf Pathname.glob HOMEBREW_CELLAR/"testball*"
+    FileUtils.rm_rf Pathname.glob HOMEBREW_CACHE/"testball*"
+    FileUtils.rm_rf Pathname.glob HOMEBREW_LOCK_DIR/"testball*"
```
I'm wondering if you can just do e.g. `FileUtils.rm_rf HOMEBREW_LOCK_DIR` and skip the glob?

MikeMcQuaid replied:
I've taken this approach in https://github.com/Homebrew/brew/pull/475.

MikeMcQuaid replied:
@eirinikos Now #475 is merged you can forget about ensure blocks in future 😀

MikeMcQuaid commented on Library/Homebrew/test/test_integration_cmds.rb:
```diff
@@ -719,4 +719,43 @@ def test_services
+    FileUtils.rm_rf Pathname.glob HOMEBREW_CACHE/"testball*"
+    FileUtils.rm_rf Pathname.glob HOMEBREW_LOCK_DIR/"testball*"
+    FileUtils.rm_rf HOMEBREW_LIBRARY/"LinkedKegs"
+    FileUtils.rm_rf HOMEBREW_PREFIX/"opt"
```
There's still a leak unfortunately 😭 http://bot.brew.sh/job/Brew%20Pull%20Requests/869/version=el_capitan/testReport/junit/brew-test-bot/el_capitan/tests___official_cmd_taps/

MikeMcQuaid replied:
Can ignore this now #475 is merged.

eirinikos commented:
I've removed the `ensure` blocks thanks to [#475](https://github.com/Homebrew/brew/pull/475). 🎉 

Are any other revisions needed at this point? Wondering if I've missed any details in other pull requests that have since been merged...

UniqMartin commented on Library/Homebrew/test/test_integration_cmds.rb:
```diff
@@ -110,11 +110,20 @@ def setup_test_formula(name, content = nil)
     formula_path = CoreTap.new.formula_dir/"#{name}.rb"
 
     case name
-    when "testball"
```
I wonder if you could, instead of moving this case to the bottom, just modify this to `/^testball/` so that it matches all formula names that start with “testball”. This would keep the diff smaller and we could more easily add additional cases in the future (if needed) without causing any name conflict.

UniqMartin commented on Library/Homebrew/test/test_integration_cmds.rb:
```diff
@@ -110,11 +110,20 @@ def setup_test_formula(name, content = nil)
         desc "Some test"
         homepage "https://example.com/#{name}"
-        url "file://#{File.expand_path("..", __FILE__)}/tarballs/#{name}-0.1.tbz"
+        url "file://#{File.expand_path("..", __FILE__)}/tarballs/testball-0.1.tbz"
```
If you're happy with my `/^testball/` suggestion, this would of course need to stay modified.

UniqMartin commented:
> Are any other revisions needed at this point? Wondering if I've missed any details in other pull requests that have since been merged...

One minor suggestion, but apart from that I think we're good. Haven't spotted anything that would be missing or that would require adjustments to changes that have happened recently.

eirinikos commented:
Thanks @UniqMartin for the `/^testball/` suggestion! It makes sense to me. 👍

UniqMartin commented:
Thanks again @eirinikos. The PR has been in the loop for quite a while and has undergone quite some changes (and even sparked some fundamental adjustments to our integration tests). What a journey—despite the seemingly trivial addition. Good work everyone!

MikeMcQuaid commented:
🎉 @eirinikos!
